python/lsst/daf/butler/datastores/s3Datastore.py

```diff
@@ -0,0 +1,719 @@
+        lengths = {"path": 256, "formatter": 128, "storage_class": 64,
+                   "checksum": 128}
+        self.records = DatabaseDict.fromConfig(self.config["records"], types=types,
+                                               value=self.RecordTuple, key="dataset_id",
+                                               lengths=lengths, registry=registry)
+
+    def __str__(self):
+        return self.root
+
+    def addStoredFileInfo(self, ref, info):
+        """Record internal storage information associated with this
+        `DatasetRef`
+        Parameters
+        ----------
+        ref : `DatasetRef`
+            The Dataset that has been stored.
+        info : `StoredFileInfo`
+            Metadata associated with the stored Dataset.
+        """
+        self.records[ref.id] = self.RecordTuple(formatter=info.formatter, path=info.path,
+                                                storage_class=info.storageClass.name,
+                                                checksum=info.checksum, file_size=info.size)
+
+    def removeStoredFileInfo(self, ref):
+        """Remove information about the file associated with this dataset.
+        Parameters
+        ----------
+        ref : `DatasetRef`
+            The Dataset that has been removed.
+        """
+        del self.records[ref.id]
+
+    def getStoredFileInfo(self, ref):
+        """Retrieve information associated with file stored in this
+        `Datastore`.
+        Parameters
+        ----------
+        ref : `DatasetRef`
+            The Dataset that is to be queried.
+        Returns
+        -------
+        info : `StoredFileInfo`
+            Stored information about this file and its formatter.
+        Raises
+        ------
+        KeyError
+            Dataset with that id can not be found.
+        """
+        record = self.records.get(ref.id, None)
+        if record is None:
+            raise KeyError("Unable to retrieve formatter associated with Dataset {}".format(ref.id))
+        # Convert name of StorageClass to instance
+        storageClass = self.storageClassFactory.getStorageClass(record.storage_class)
+        return StoredFileInfo(record.formatter, record.path, storageClass,
+                              checksum=record.checksum, size=record.file_size)
+
+    def exists(self, ref):
+        """Check if the dataset exists in the datastore.
+
+        Parameters
+        ----------
+        ref : `DatasetRef`
+            Reference to the required dataset.
+
+        Returns
+        -------
+        exists : `bool`
+            `True` if the entity exists in the `Datastore`.
+        """
+        # Get the file information (this will fail if no file)
+        try:
+            storedFileInfo = self.getStoredFileInfo(ref)
+        except KeyError:
+            return False
+
+        loc = self.s3locationFactory.fromPath(storedFileInfo.path)
+        return s3CheckFileExists(self.client, loc.bucket, loc.path)[0]
```

Comment:
I think s3CheckFileExists should be taking a `ButlerURI` once I add that.

Reply:
I changed the `s3CheckFileExists` to accept a `Location`, `ButlerURI` or specific `(bucket, filepath)` pair. 

I've left the old way in place because in ingest at some point we could be ingesting a local file. Before we actually do the upload we need to check if a file of the same name exists in the Bucket already so we don't overwrite it. Then we use destination location for bucket name, and source location to produce a predicted path in bucket. 
Also at some time later after upload we check if file actually exists in the bucket. I believe its good to construct the path from mix of the source and destination locations because it makes me feel more secure in the fact that the file actually got uploaded in the correct location in the Bucket and that no mishaps happened.

(Will be up in minutes).